### ISSUE TITLE
[BUG] Resource wildwebmidi.data not found in Composer-based installations

### ISSUE THREAD

**Issue by markusweigelt** [kitodo/kitodo-presentation#1367]
## Description

`wildwebmidi.js` references an incorrect file location.

A significant change in TYPO3 12 is to load public resources from the `_assets` folder, where the public resource folders of extensions are symlinked.

[A clear and concise description of what the bug is.](https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/12.0/Important-98484-ExtensionsOutsideOfDocumentRootForComposer-basedTYPO3Installations.html)

## Reproduction

Steps to reproduce the behaviour:

1. Install `dlf` extension in Composer-based Typo3 12 installation

## Expected Behavior

`wildwebmidi.data` loads without 404 from `_assets` folder

## Solution

Use path of caller file `wildwebmidi.js` maybe with `document.currentScript`.

Because it looks like a library that is also used in other projects, I didn't fix it.

**Comment by markusweigelt** [comment]
Does not work with `document.currentScript.src` because the current script is a merged version of all JavaScript files and is located in a different directory than the extension's assets.

Since I was interested in finding a solution, I discovered a satisfactory approach that works in both and next TYPO3 versions out of the box.

```
[typo3.branch=="11.5"]
  page.inlineSettings {
    dlfAssets = /typo3conf/ext/dlf/Resources/Public/
  }
[else]
  page.inlineSettings {
    dlfAssets = /_assets/007831ab53eafe3f6be2b9b1baf2a8d6/
  }
[end]
```
https://github.com/markusweigelt/kitodo-presentation/blob/c2534ae53a65c8f95a8d81db234b2038eaa4ae40/Configuration/TypoScript/setup.typoscript#L72-L80

After adding this to TypoScript and clearing the cache, the URL can be correctly built with this inline setting.

`var REMOTE_PACKAGE_BASE= TYPO3.settings.TS.dlfAssets+"JavaScript/WildWebMidi/wildwebmidi.data";`

https://github.com/markusweigelt/kitodo-presentation/blob/c2534ae53a65c8f95a8d81db234b2038eaa4ae40/Resources/Public/JavaScript/WildWebMidi/wildwebmidi.js#L4

**Comment by sebastian-meyer** [comment]
This won't work consistently, because the asset hash is not the same for every installation. It is derived from the full path and that depends on where TYPO3 is installed. We need to find another way to fix this.

**Comment by sebastian-meyer** [comment]
Assigning @chrizzor and @thomaslow, because the former introduced the MIDI player while the latter was responsible for TYPO3 v12 compatibility. Maybe you can figure out an approbiate solution together?

**Comment by markusweigelt** [comment]
@sebastian-meyer 
> This won't work consistently, because the asset hash is not the same for every installation. It is derived from the full path and that depends on where TYPO3 is installed. We need to find another way to fix this.

Are you sure? As I understand the code, it’s the relative path from which the MD5 is generated. It’s been a while since I tested it. 
As far as I can remember, I had checked it. 
https://github.com/TYPO3/typo3/blob/052b80b2d0b49bb6c3078e943b4a2ba1babd512b/typo3/sysext/core/Classes/Utility/PathUtility.php#L71

**Comment by sebastian-meyer** [comment]
Yes, maybe I am wrong here. This should be tested for different scenarios.

**Comment by thomaslow** [comment]
I had a similar problem in the `kitodo-demo` theme. I used the following snippet to get the base asset path exposed in Javascript, see [branch](https://github.com/kitodo/kitodo-demo/compare/master...thomaslow:kitodo-demo:typo3-v12-support):

```
<f:asset.script identifier="js-resource-path-inline-script" priority="true">
    const EXT_PRESENTATION_PACKAGE_RESOURCE_PATH = "{f:uri.resource(path:'EXT:presentation_package/Resources/Public')}";
</f:asset.script>
```

I think that worked both for Typo3 v11 and v12 and returned the respective old and new asset path.

Not sure whether it helps in this case.

**Comment by markusweigelt** [comment]
A practical solution, but I don't think it's suitable for general use, as this adjustment would need to be made in every layout.

I had another idea, which is to render a meta tag, e.g., `dlf:resourcepath`, with the public resource path of `dlf` extension on the page using the AbstractController.

```
        $registry = GeneralUtility::makeInstance(MetaTagManagerRegistry::class);
        $metaTagManager = $registry->getManagerForProperty('dlf:resourcePath');
        if (empty($metaTagManager->getProperty('dlf:resourcePath'))) {
            $metaTagManager->addProperty('dlf:resourcePath', PathUtility::getPublicResourceWebPath('EXT:dlf/Resources/Public'));
        }

```

Following tag will then appear in the HTML head:

`<meta name="dlf:resourcepath" content="/typo3conf/ext/dlf/Resources/Public">`

which can then be used through the following JavaScript call.

`document.querySelector("meta[name='dlf:resourcepath']").getAttribute("content")`

@sebastian-meyer, @thomaslow  What do you think of this solution?

**Comment by sebastian-meyer** [comment]
The [official documentation](https://docs.typo3.org/m/typo3/reference-coreapi/main/en-us/Administration/Upgrade/MigrateToComposer/AssetMigration.html) has some tips on how to access the public resource path from different environments: 

> References within the same extension should use relative links, for example use `background-image: url('../Images/logo.jpg')` instead of `background-image: url('/typo3conf/ext/my_extension/Resources/Public/Images/logo.jpg')`.

> You can use TypoScript/PHP/Fluid as mentioned above to create variables with resolved asset URI locations. These variables can utilize the `EXT:my_extension/Resources/Public/...` notation, and can be passed along to a JavaScript variable or a HTML DOM/data attribute, so it can be further evaluated.

If we can't use relative links your proposal of adding the resolved path into the DOM is fine, but I'd recommend using a `data-path` attribute instead of a `meta` tag (see also [here](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/data-*)).

**Comment by markusweigelt** [comment]
Exactly, I had that documentation in mind as well. Unfortunately, relative paths don’t work because compressing and merging all JavaScript files causes the path to the asset folder of the respective extension to be lost.

I have been looking for a general approach where the path for all JavaScript files is available. I will implement your solution to fix the problem, and if the path is needed elsewhere, we can generalize the implementation later.
